diff --git a/binding.gyp b/binding.gyp
index 5f63978b07ab50aaf7523219a2170ec737a6b5db..1f77c45adf60c9fa532775e43a06d507a0342e8f 100644
--- a/binding.gyp
+++ b/binding.gyp
@@ -6,8 +6,12 @@
     'conditions': [
       ['OS=="win"', {
         'msvs_configuration_attributes': {
+          'CharacterSet': '0',
           'SpectreMitigation': 'Spectre'
         },
+        'defines': [
+          'NOMINMAX'
+        ],
         'msvs_settings': {
             'VCCLCompilerTool': {
               'AdditionalOptions': [
diff --git a/deps/winpty/src/gen/GenVersion.h b/deps/winpty/src/gen/GenVersion.h
new file mode 100644
index 0000000000000000000000000000000000000000..24cb8367847325c69bc45ce1eb8e0a855d3e5b27
--- /dev/null
+++ b/deps/winpty/src/gen/GenVersion.h
@@ -0,0 +1,3 @@
+// AUTO-GENERATED - hardcoded for node-pty build
+const char GenVersion_Version[] = "0.4.4-dev";
+const char GenVersion_Commit[] = "none";
diff --git a/deps/winpty/src/winpty.gyp b/deps/winpty/src/winpty.gyp
index 1ac5758bedd8cf54f32280dea4e4aeb5afdee30d..443da22beead7317a0fa1fd86fc88cd089437ccd 100644
--- a/deps/winpty/src/winpty.gyp
+++ b/deps/winpty/src/winpty.gyp
@@ -10,7 +10,7 @@
     #    make -j4 CXX=i686-w64-mingw32-g++ LDFLAGS="-static -static-libgcc -static-libstdc++"
 
     'variables': {
-        'WINPTY_COMMIT_HASH%': '<!(cmd /c "cd shared && GetCommitHash.bat")',
+        'WINPTY_COMMIT_HASH%': 'none',
     },
     'target_defaults' : {
         'defines' : [
@@ -20,9 +20,7 @@
             'NOMINMAX',
         ],
         'include_dirs': [
-            # Add the 'src/gen' directory to the include path and force gyp to
-            # run the script (re)generating the version header.
-            '<!(cmd /c "cd shared && UpdateGenVersion.bat <(WINPTY_COMMIT_HASH)")',
+            'gen',
         ]
     },
     'targets' : [
diff --git a/src/win/conpty.cc b/src/win/conpty.cc
index 7b286d3d644c26141df516929703aa6e129df4b2..b900b4d6875725dcc2fbd98ef72176d2202d9112 100644
--- a/src/win/conpty.cc
+++ b/src/win/conpty.cc
@@ -25,17 +25,34 @@
 #include "conpty.h"
 
 // Taken from the RS5 Windows SDK, but redefined here in case we're targeting <= 17134
-#ifndef PROC_THREAD_ATTRIBUTE_PSEUDOCONSOLE
-#define PROC_THREAD_ATTRIBUTE_PSEUDOCONSOLE \
-  ProcThreadAttributeValue(22, FALSE, TRUE, FALSE)
-
+#ifndef HPCON
 typedef VOID* HPCON;
+#endif
+
+#ifndef PFNCREATEPSEUDOCONSOLE
+#define PFNCREATEPSEUDOCONSOLE PFNCREATEPSEUDOCONSOLE_PTY
 typedef HRESULT (__stdcall *PFNCREATEPSEUDOCONSOLE)(COORD c, HANDLE hIn, HANDLE hOut, DWORD dwFlags, HPCON* phpcon);
+#endif
+#ifndef PFNRESIZEPSEUDOCONSOLE
+#define PFNRESIZEPSEUDOCONSOLE PFNRESIZEPSEUDOCONSOLE_PTY
 typedef HRESULT (__stdcall *PFNRESIZEPSEUDOCONSOLE)(HPCON hpc, COORD newSize);
+#endif
+#ifndef PFNCLEARPSEUDOCONSOLE
+#define PFNCLEARPSEUDOCONSOLE PFNCLEARPSEUDOCONSOLE_PTY
 typedef HRESULT (__stdcall *PFNCLEARPSEUDOCONSOLE)(HPCON hpc);
+#endif
+#ifndef PFNCLOSEPSEUDOCONSOLE
+#define PFNCLOSEPSEUDOCONSOLE PFNCLOSEPSEUDOCONSOLE_PTY
 typedef void (__stdcall *PFNCLOSEPSEUDOCONSOLE)(HPCON hpc);
+#endif
+#ifndef PFNRELEASEPSEUDOCONSOLE
+#define PFNRELEASEPSEUDOCONSOLE PFNRELEASEPSEUDOCONSOLE_PTY
 typedef void (__stdcall *PFNRELEASEPSEUDOCONSOLE)(HPCON hpc);
+#endif
 
+#ifndef PROC_THREAD_ATTRIBUTE_PSEUDOCONSOLE
+#define PROC_THREAD_ATTRIBUTE_PSEUDOCONSOLE \
+  ProcThreadAttributeValue(22, FALSE, TRUE, FALSE)
 #endif
 
 struct pty_baton {
